src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -0,0 +1,195 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.SamRecordClipper.ClippingMode
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{DelimitedDataParser, Io}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools.SamPairUtil.PairOrientation
+import htsjdk.samtools._
+import htsjdk.samtools.util._
+import Math.abs
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+
+@clp(group=ClpGroups.SamOrBam, description=
+  """
+    |Trims primers from reads post-alignment.  Takes in a BAM file of aligned reads
+    |and a tab-delimited file with five columns (chrom, left_start, left_stop, right_start,
+    |and right_stop) which provide the 1-based inclusive start and end positions of the
+    |primers for each amplicon.
+    |
+    |Paired end reads that map to a given amplicon position are trimmed so that the
+    |alignment no-longer includes the primer sequences.
+    |
+    |All other aligned reads have the maximum primer length trimmed!
+  """)
+class TrimPrimers
+( @arg(flag="i", doc="Input BAM file.")  val input: PathToBam,
+  @arg(flag="o", doc="Output BAM file.") val output: PathToBam,
+  @arg(flag="p", doc="File with primer locations.") val primers: FilePath,
+  @arg(flag="H", doc="If true, hard clip reads, else soft clip.") val hardClip: Boolean = false,
+  @arg(flag="S", doc="Match to primer locations +/- this many bases.") val slop: Int = 5,
+  @arg(flag="s", doc="Sort order of output BAM file (defaults to input sort order).") val sortOrder: Option[SortOrder] = None
+)extends FgBioTool with LazyLogging {
+  private val mode = if (hardClip) ClippingMode.Hard else ClippingMode.Soft
+
+  Io.assertReadable(input)
+  Io.assertReadable(primers)
+  Io.assertCanWriteFile(output)
+
+  /** A Locatable Amplicon class. */
+  private case class Amplicon(chrom: String, leftStart: Int, leftEnd: Int, rightStart: Int, rightEnd: Int) extends Locatable {
+    def leftPrimerLength: Int    = CoordMath.getLength(leftStart, leftEnd)
+    def rightPrimerLength: Int   = CoordMath.getLength(rightStart, rightEnd)
+    def longestPrimerLength: Int = Math.max(leftPrimerLength, rightPrimerLength)
+
+    override def getContig: String = chrom
+    override def getStart: Int = leftStart
+    override def getEnd: Int = rightEnd
+  }
+
+  override def execute(): Unit = {
+    val in = SamReaderFactory.make().open(input)
+    val outSortOrder = sortOrder.getOrElse(in.getFileHeader.getSortOrder)
+    val outHeader = in.getFileHeader.clone()
+    outHeader.setSortOrder(outSortOrder)
+    val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
```

Comment:
It's because I'm iterating over the input in queryname order (either because it is, or because I sorted it), so when writing the output I need to re-sort if the desired sort isn't queryname.